libraries/chain/webassembly/eos-vm-oc/executor.cpp

```diff
@@ -0,0 +1,251 @@
+
+   //as a double check that the control block pointer is what we expect, look for the magic
+   if(cb_in_main_segment->magic != signal_sentinel)
+      goto notus;
+
+   //was wasm running? If not, this SEGV was not due to us
+   if(cb_in_main_segment->is_running == false)
+      goto notus;
+
+   //was the segfault within code?
+   if((uintptr_t)info->si_addr >= cb_in_main_segment->execution_thread_code_start &&
+      (uintptr_t)info->si_addr < cb_in_main_segment->execution_thread_code_start+cb_in_main_segment->execution_thread_code_length)
+         siglongjmp(*cb_in_main_segment->jmp, EOSVMOC_EXIT_CHECKTIME_FAIL);
+
+   //was the segfault within data?
+   if((uintptr_t)info->si_addr >= cb_in_main_segment->execution_thread_memory_start &&
+      (uintptr_t)info->si_addr < cb_in_main_segment->execution_thread_memory_start+cb_in_main_segment->execution_thread_memory_length)
+         siglongjmp(*cb_in_main_segment->jmp, EOSVMOC_EXIT_SEGV);
+
+notus:
+   if(chained_handler)
+      chained_handler(sig, info, ctx);
+   ::signal(sig, SIG_DFL);
+   ::raise(sig);
+   __builtin_unreachable();
+}
+
+static int32_t grow_memory(int32_t grow, int32_t max) {
+   EOSVMOC_MEMORY_PTR_cb_ptr;
+   uint64_t previous_page_count = cb_ptr->current_linear_memory_pages;
+   int32_t grow_amount = grow;
+   uint64_t max_pages = max;
+   if(grow == 0)
+      return (int32_t)cb_ptr->current_linear_memory_pages;
+   if(previous_page_count + grow_amount > max_pages)
+      return (int32_t)-1;
+
+   uint64_t current_gs;
+   arch_prctl(ARCH_GET_GS, &current_gs);
+   current_gs += grow_amount * memory::stride;
+   arch_prctl(ARCH_SET_GS, (unsigned long*)current_gs);
+   cb_ptr->current_linear_memory_pages += grow_amount;
+
+   if(grow_amount > 0)
+      memset((void*)(cb_ptr->full_linear_memory_start + previous_page_count*64u*1024u), 0, grow_amount*64u*1024u);
+
+   return (int32_t)previous_page_count;
+}
+static intrinsic grow_memory_intrinsic EOSVMOC_INTRINSIC_INIT_PRIORITY("eosvmoc_internal.grow_memory", IR::FunctionType::get(IR::ResultType::i32,{IR::ValueType::i32,IR::ValueType::i32}), (void*)&grow_memory,
+  boost::hana::index_if(intrinsic_table, ::boost::hana::equal.to(BOOST_HANA_STRING("eosvmoc_internal.grow_memory"))).value()
+);
+
+//This is effectively overriding the eosio_exit intrinsic in wasm_interface
+static void eosio_exit(int32_t code) {
+   EOSVMOC_MEMORY_PTR_cb_ptr;
+   siglongjmp(*cb_ptr->jmp, EOSVMOC_EXIT_CLEAN_EXIT);
+   __builtin_unreachable();
+}
+static intrinsic eosio_exit_intrinsic("env.eosio_exit", IR::FunctionType::get(IR::ResultType::none,{IR::ValueType::i32}), (void*)&eosio_exit,
+  boost::hana::index_if(intrinsic_table, ::boost::hana::equal.to(BOOST_HANA_STRING("env.eosio_exit"))).value()
+);
+
+static void throw_internal_exception(const std::string& s) {
+   EOSVMOC_MEMORY_PTR_cb_ptr;
+   *cb_ptr->eptr = std::make_exception_ptr(wasm_execution_error(FC_LOG_MESSAGE(error, s)));
+   siglongjmp(*cb_ptr->jmp, EOSVMOC_EXIT_EXCEPTION);
```

Comment:
This leaks s.  It may or may not actually leak memory depending on how large std::string's SBO buffer is.